### ISSUE TITLE
[BUG] Docs say es6 modules can't export class objects, but that's not true.

### ISSUE THREAD

**Issue by trusktr** [Microsoft/TypeScript#13627]
The docs [say](https://www.typescriptlang.org/docs/handbook/declaration-files/templates/module-class-d-ts.html) that

> ```
> /*~ Note that ES6 modules cannot directly export class objects.
> ```

but modules can export classes, as in

```ts
export default
class Foo { /*...*/ }
```

**Comment by blakeembrey** [comment]
I believe that, in context, this is alluding to the fact that ES6 modules do not have a way to mimic CommonJS `module.exports =` style modules (e.g. `export =` in TypeScript). Agree the wording is confusing though and should probably be something like:

> Note that ES6 module syntax can not be used to describe CommonJS accurately. Instead, you must use `export =` to export the class and `import x = require('someLibrary')` to import it.

**Comment by mhegazy** [comment]
In commonjs you can replace the module object with a class or a function, this is not possible in ES6. The default export is just a property export that has the name "default".

**Comment by blakeembrey** [comment]
@mhegazy Agreed, but I can see it's confusing for anyone new to ES6 modules or TypeScript and could be clearer. For instance, as far as anyone using ES6 modules cares, `export default` is "directly" exported. You can see this through the amount of confused people over `import x from` and `import x =`.

**Comment by mhegazy** [comment]
A PR clarifying the text would be welcomed.

**Comment by djindjic** [comment]
Yes, I am also trying to understand this part and it's not clear to me. I also don't understand @blakeembrey 's `ES6 modules do not have a way to mimic CommonJS`, why should they have to mimic commonjs in context of this doc? Maybe to clarify difference between `import x from` and `import x =`?

**Comment by blakeembrey** [comment]
@djindjic I don't know how someone gets to that page to answer, but I am assuming that document is for helping people write a definition for a CommonJS module. You can't type CommonJS modules that use `module.exports =` using the ES6 syntax, you must use `export =` in TypeScript.

`import x = require('')` is the same as `var x = require('')`. `import x from ''` is part of the ES6 spec and imports the `default` export. There's no such thing as `module.exports` in ES6 modules. Looking at the transpiled exports probably makes the difference easier to understand.

* `export default` becomes `exports.default` and is imported using `import anyNameHere`
* `export { foo }` becomes `exports.foo` and is imported using `import { foo }`

**Comment by djindjic** [comment]
Maybe I have different perspective, but I just want to learn how to write a definition for commonjs module and I didn't have idea to write module.export in definition file. By my opinion this text should be removed or I cant catch why is it important in context of writing definitions.

Can you help me to find some hello world example with real commonjs lib file, definition and usage of it in ex. main.ts?

**Comment by blakeembrey** [comment]
@djindjic You use `export =` to mimic `module.exports` for CommonJS. You can check out https://github.com/typings/typings/blob/master/docs/examples.md#external-module-with-export- for some examples of CommonJS libraries typed.

**Comment by blakeembrey** [comment]
The distinction is important, though the existing message is really confusing, because a lot of people don't notice that `export =` and `export default` is much different, just as `import foo =` and import foo from` is also different. For instance, I've made a few PRs like https://github.com/facebook/dataloader/pull/51 fixing these sorts of mishaps where matching the export correctly wasn't considered when the definition was written.

**Comment by djindjic** [comment]
thanks for examples @blakeembrey! I also found [this](https://github.com/angular/angular-cli/blob/092e67315344f800c9e01f5324d649394b6c847c/docs/documentation/stories/third-party-lib.md), and whan I write:
`import * as typelessPackage from 'typeless-package';`
is it actually pointing to definition or real library? It's totally confusing for me, don't know where I can learn about this.

Thanks again for you help.

**Comment by blakeembrey** [comment]
@djindjic When you do any imports in TypeScript, it's always using a definition to verify that it's correct. It's only when you're executing JavaScript that it's using the library. Now, TypeScript has different ways of resolving that definition - which may be from a `.d.ts` that's bundled, a different `.d.ts` included with `declare module ''`, the `@types` NPM namespace or even a JavaScript file when `allowJs: true`. 

For the specific syntax, `import * as`, it's just an ES6 import that imports _all_ the exported properties from the definition. You can check out the ES6 module spec for more specifics, but basically if you had `import * as x` from a module that `export { foo, bar }`, you'd be able to access `x.foo` and `x.bar`. In this respect, it can be used to import properties from CommonJS if you're transpiling your module to CommonJS (since `import { foo } from 'x'` translates to `var foo = require('x').foo`).

**Comment by djindjic** [comment]
thanks @blakeembrey, please give me few more answers because I cant find out it in docs.
By the (link)[https://github.com/typings/typings/blob/master/docs/examples.md] you provided, I wanted to write definition for commonjs module. Example (number 3)[https://github.com/typings/typings/blob/master/docs/examples.md#external-module-with-export-] fits it (I think) i pushed mu code to this (repo)[https://github.com/djindjic/tsc-playground/tree/b82296d36379b11a141fb6bafad66a7dc90ed93d]. There is only 3 files: comnonjs module, definition and main.ts which use it. I run `tsc main.ts` and get this error:
>main.ts(1,26): error TS2497: Module '"/Users/aleksandar/projects/tsc-playground/domready"' resolves to a non-module entity and cannot be imported using this construct

This is to basic case, but I cant find in documentation how to do it and understand how tsc works.
